mantle/kola/harness.go

```diff
@@ -844,24 +849,47 @@ func metadataFromTestBinary(executable string) (*externalTestMeta, error) {
 	defer f.Close()
 	r := bufio.NewReader(io.LimitReader(f, 8192))
 	meta := &externalTestMeta{Exclusive: true}
+	inmeta := false    // true if we saw a ## kola: prefix after which we expect YAML
+	metadatayaml := "" // accumulated YAML metadata
 	for {
 		line, err := r.ReadString('\n')
 		if err == io.EOF {
 			break
 		} else if err != nil {
 			return nil, err
 		}
-		if !strings.HasPrefix(line, InstalledTestMetaPrefix) {
+		if strings.HasPrefix(line, InstalledTestMetaPrefixYaml) {
+			if inmeta {
+				return nil, fmt.Errorf("found multiple %s", InstalledTestMetaPrefixYaml)
+			}
+			inmeta = true
 			continue
 		}
-		buf := strings.TrimSpace(line[len(InstalledTestMetaPrefix):])
-		dec := json.NewDecoder(strings.NewReader(buf))
-		dec.DisallowUnknownFields()
-		meta = &externalTestMeta{Exclusive: true}
-		if err := dec.Decode(meta); err != nil {
-			return nil, errors.Wrapf(err, "parsing %s", line)
+		if inmeta {
+			if !strings.HasPrefix(line, "## ") {
+				if err := yaml.UnmarshalStrict([]byte(metadatayaml), &meta); err != nil {
+					return nil, err
+				}
+				break
+			} else {
+				metadatayaml += fmt.Sprintf("%s\n", line[3:])
+			}
+			continue
+		}
+		// Otherwise, we look for the legacy JSON metadata
+		if strings.HasPrefix(line, InstalledTestMetaPrefix) {
+			if inmeta {
+				return nil, fmt.Errorf("found both yaml and json test prefixes (%v %v)", InstalledTestMetaPrefixYaml, InstalledTestMetaPrefix)
+			}
```

Comment:
```suggestion
```

Did you intend to add this in your latest upload?

It's dead code IIUC because `if inmeta {` we'll go into the conditional logic above, which either does a `break` or `continue`.

Reply:
Good catch; I moved the json bits up which actually allowed us to drop the `continue` statements and simplify the overall control flow.